pkg/action/executor/change_request_controller.go

```diff
@@ -0,0 +1,206 @@
+	}
+	err = unstructured.SetNestedField(cR.Object, repo, "spec", "source")
+	if err != nil {
+		return err
+	}
+	err = unstructured.SetNestedField(cR.Object, path, "spec", "path")
+	if err != nil {
+		return err
+	}
+	err = unstructured.SetNestedField(cR.Object, revision, "spec", "branch")
+	if err != nil {
+		return err
+	}
+
+	// The secret name and namspace if listed override the ones configured on the CR controller
+	annotations := cr.obj.GetAnnotations()
+	secretName, exists := annotations[GitopsSecretNameAnnotationKey]
+	if exists {
+		if err = unstructured.SetNestedField(cR.Object, secretName, "spec", "secretRef", "name"); err != nil {
+			return err
+		}
+	}
+	secretNamespace, exists := annotations[GitopsSecretNamespaceAnnotationKey]
+	if exists {
+		if err = unstructured.SetNestedField(cR.Object, secretNamespace, "spec", "secretRef", "namespace"); err != nil {
+			return err
+		}
+	}
+
+	var patches []interface{}
+	patches = append(patches, map[string]interface{}{
+		"op":    "replace",
+		"path":  "/spec/replicas",
+		"value": replicas,
+	})
+	patches = append(patches, map[string]interface{}{
+		"op":    "replace",
+		"path":  "/spec/template/spec",
+		"value": podSpec,
+	})
+
+	err = unstructured.SetNestedSlice(cR.Object, patches, "spec", "patchItems")
+	if err != nil {
+		return err
+	}
+
+	res := schema.GroupVersionResource{
+		Group:    CRGroup,
+		Version:  CRVersion,
+		Resource: CRResourceName,
+	}
+	namespacedClient := cr.dynClient.Resource(res).Namespace(cr.obj.GetNamespace())
+
+	// We try to create a new CR resource with the update info.
+	// It might be that this action has been executed once and we are trying it
+	// again in which case we try to only update the existing resource
+	// Also we create the ChangeRequest resource with same name as workload obj
+	_, outerErr := namespacedClient.Create(context.Background(), cR, metav1.CreateOptions{})
+	if apierrors.IsAlreadyExists(outerErr) {
+		cRSpec, found, err := unstructured.NestedMap(cR.Object, "spec")
+		if err != nil || !found {
+			return fmt.Errorf("could not retrieve spec fields from ChangeRequest: %s: %v", cR.GetName(), err)
+		}
+
+		existing, err := namespacedClient.Get(context.Background(), cR.GetName(), metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
+
+		err = unstructured.SetNestedMap(existing.Object, cRSpec, "spec")
+		if err != nil {
+			return err
+		}
+
+		// TODO: Implement a wait loop here to retrieve the CR again and wait
+		// until the status is either "Complete" or "Failed".
+		// Once the retrieved status is complete, resetting the status to ""
+		// and updating the spec would mean the CR controller will try to push
+		// the changes again. We will also need to ensure that pushing the change
+		// does not mean creating multiple PRs if the previous one exists.
+
+		// TODO: Reset the status too, because if the status still remains
+		// 'completed' the change-reconciler would not do anything.
+		// Brainstorm a little more on this.
```

Comment:
It's tricky to synchronize on a mutable object.  I wonder if it's cleaner to create a new change request each time with a timestamp.  The change reconciler will populate the status as complete, failed or skipped (if a newer one exists before the older one is processed).  The change request controller can choose to delete a change request once it's fulfilled.

Reply:
Removed the file as in [this comment](https://github.com/turbonomic/kubeturbo/pull/655#discussion_r847272590)